---
title: "Bladder.Affymetrix.2018"
author: "Robert W. Murdoch"
date: "October 29, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is a full pipeline for importaing affymetrix .CEL data (human microarray chip data)
and running full pairwise differential abundance tests.  A full annotation table is also
referenced in order to link probe names to gene names.  

The .CEL data is read and analyzed by the "limma" package

## initializing limma and dependent packages and importing the annotation table

First, the necessary libraries are activated and the annotation table 
(downloaded from an Affymetrix website) is read in:

```{r, echo=FALSE, results="hide", message=FALSE}
library(limma)
library(statmod)
library(affy)
library(affycoretools)
library(oligo)
library(hta20transcriptcluster.db)
```
```{r}
Annotation <- read.csv("HTA-2_0.na35.2.hg19.transcript.csv/HTA-2_0.na35.2.hg19.transcript.csv", header=TRUE, skip=19)
```

The annotation table is very large and many columns are merged redundant database references (using "//" as separator). it is far too wide to display here!
However, it is very powerful as it links the probes to a huge number of databases.

##Mapping probe ID to genes and pathways
The Annotation table has several columns:
```{r}
colnames(Annotation)
```

Many of these are potentially very useful.  For now, I am useing the gene_assignment column.  This contains many names separated by //.  A single entry can look like this:

```{r}
Annotation[1,"gene_assignment"]
```

The second one is a gene name that links well to many databases and the third provides a description. 
The gene names can be fed into the Reactome online database for pathway mapping.
We will create a focused mapping table that only has the necessary fields. This object
will be used to attach gene names and functions to the final diff.expression tables
```{r}
genenames.all <- Annotation
gene.split <- data.frame(do.call('rbind',strsplit(as.character(genenames.all$gene_assignment),"//",fixed=T)))
probe.gene.map <- data.frame(gene.split[,c(2,3)])
probe.gene.map$probe <- Annotation[,1]
probe.gene.map <- probe.gene.map[,c(3,1,2)]
probe.gene.map$X2 <- gsub("[[:space:]]", "", probe.gene.map$X2)
colnames(probe.gene.map) <- c("probe","name","description")

head(probe.gene.map)
```

## Test matrices

Differential abundance can only be performed on pairs of sample sets.  Each pairwise test has to be configured as a separate file.
Importing the test matrices:

```{r}
targets <- read.csv("targets.txt", as.is=T)
targets.tGF.cGF <- read.csv("targets.GFonly.txt", as.is=T)
targets.tGFR.cGFR <- read.csv("targets.GFRonly.txt", as.is=T)
targets.cGFR.cGF <- read.csv("targets.noDrug.Strain.txt", as.is=T)
targets.tGFR.tGF <- read.csv("targets.drugonly.txt", as.is=T)
```

all of the samples and metadata are listed here:

```{r}
targets
```

The matrices are set up as such; this matrix is for comparing non-treated and treated samples within in strain GF (control strain). Note thta this is just one of the tests; we now have matrices for 3 other test schemes.

```{r}
targets.tGF.cGF
```

## Reading the .CEL files and normalizing.
* after being read in, the project is a complex object, not directly viewable.
* the "exprs" function from the Biobase package pulls out the expression data as a table; at the same time, control probes are removed.
* this code chunk also writes the full normalized results as a .tsv file (although there are no stats or comparisons)

```{r}
dat <- read.celfiles(targets$FileName) #read .CEL files
project.bgcorrect.norm.avg <- rma(dat, background=TRUE, normalize=TRUE) #normalize
#this may remove control probes, it should only include probes which actually link to genes
project.bgcorrect.norm.avg.filt <- exprs(project.bgcorrect.norm.avg)[which(rownames(project.bgcorrect.norm.avg) %in% Annotation[,1]),]
#this writes the full results as a file
write.table(project.bgcorrect.norm.avg.filt, "full.expression.results.tsv", sep="\t", row.names = TRUE, quote=FALSE)                                        
```

## The expression table

After export and filtration, the simplified table looks like 
(this is the first 6 probes and their normalized intensities):
```{r}
head(data.frame(project.bgcorrect.norm.avg.filt))
```



# Generating expression tables for differential abundance tests

This full data table that we just generated is only for record keeping purposes only.  For each of the pairwise tests we will do, we will repeat this except that it will be limited to the samples we are testing with:
*define the test using the target matrix table (e.g. c vs. t, GF vs GFR)
*read in the .CEL files limited to the matrix table (only a subset of .CELs are read in)
*normalize
*define your tests and make differential abundance tables

##untreated GF vs untreated GFR

###1. Define the subset of .CEL files, read them in, and normalize them
```{r}
dat <- read.celfiles(targets.cGFR.cGF$FileName)
project.bgcorrect.norm.avg <- rma(dat, background=TRUE, normalize=TRUE)
project.bgcorrect.norm.avg.filt <- exprs(project.bgcorrect.norm.avg)[which(rownames(project.bgcorrect.norm.avg) %in% Annotation[,1]),]
```

###2. Load in the metadata
```{r}
Exp <- factor(targets.cGFR.cGF$Experiment)
Strain <- factor(targets.cGFR.cGF$Strain, levels=c("GF","GFR"))
Treatment <- factor(targets.cGFR.cGF$Treatment, levels=c("control","drug"))
```

###3. Define the test as controlling for replicate ("exp") while testing effect of strain
This is important decision; these chips were ran from different experimental batches,
thus I have controlled for experiment, essentially running a paired analysis.
In an earlier iteration, I did not control for experiment and found very few differences due to the 
  high variance between experimental replicates
```{r}
design <- model.matrix(~Exp+Strain)
```

###4. Run "fit" algorithms and separate out the significantly different probes

```{r}
eset <- project.bgcorrect.norm.avg.filt
fit <- lmFit(eset, design)
fit <- eBayes(fit, trend=T, robust=T)
results <- decideTests(fit)
write.table(summary(results), "controls.summary", quote=F)
```

####This is a simple summary of the differences, only the far right column
```{r}
summary(results)
```

###5. Generate a table of the differentially expressed genes and attach gene name/function information
```{r, results="hide", message=FALSE}
library(data.table)
library(dplyr)
diff.cGFR.cGF <- topTable(fit, coef="StrainGFR", n=73)
setDT(diff.cGFR.cGF, keep.rownames = TRUE)[]
colnames(diff.cGFR.cGF)[1] <- "probe"
diff.cGFR.cGF <- dplyr::left_join(diff.cGFR.cGF,probe.gene.map,by="probe")
write.table(diff.cGFR.cGF, "diff.cGFR.cGF", quote = F)
write.table(diff.cGFR.cGF$name, "diff.cGFR.cGF.gene.list", quote = F)
```
```{r}
head(diff.cGFR.cGF)
```

This table can be regarded as a final output of the differential expression pipeline.

#TO DO: split this into up and down regulated for analysis purposes

##6. linking genes to pathways using Reactome.db
There is an API, a restful API, and some sort of R package "reactome.db" for this. Querying the database directly
https://reactome.org/PathwayBrowser/#TOOL=AT is simple though. Pasting in the list of genes found in the *.gene.list file that was written, we get access to a pathway map with a good deal of depth for exploration

```{r, echo=F}
knitr::include_graphics("reactome.cGFR.cGF.png")
```

This image can be zoomed in on the website for perhaps more information or dramatic focus.

The results can be downloaded and loaded into R:

```{r}
cGF.cGFR.reactome <- read.csv("reactome.cGFR.cGF.csv", header=T)
head(cGF.cGFR.reactome)
```

There is a column that signified which gene names linked to each pathway.

##Reactome also generates a spectacular report .pdf.  Be sure to download it and examine!

#To Do: 
*before generating dex tables for each of the tests separate the up and down regulated genes before sending to reactome






close dont delete